rust/datafusion/src/logical_plan/dfschema.rs

```diff
@@ -0,0 +1,415 @@
+            })
+            .collect();
+        match matches.len() {
+            0 => Err(DataFusionError::Plan(format!(
+                "No field named '{}.{}'",
+                relation_name, name
+            ))),
+            1 => Ok(matches[0].to_owned()),
+            _ => Err(DataFusionError::Plan(format!(
+                "Ambiguous reference to qualified field named '{}.{}'",
+                relation_name, name
+            ))),
+        }
+    }
+
+    /// Convert to an Arrow schema
+    pub fn to_arrow_schema(&self) -> SchemaRef {
+        SchemaRef::new(Schema::new(
+            self.fields.iter().map(|f| f.field.clone()).collect(),
+        ))
+    }
+}
+
+impl Display for DFSchema {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "{}",
+            self.fields
+                .iter()
+                .map(|field| field.qualified_name())
+                .collect::<Vec<String>>()
+                .join(", ")
+        )
+    }
+}
+
+/// DFField wraps an Arrow field and adds an optional qualifier
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct DFField {
+    /// Optional qualifier (usually a table or relation name)
+    qualifier: Option<String>,
+    /// Arrow field definition
+    field: Field,
+}
+
+impl DFField {
+    /// Creates a new `DFField`
+    pub fn new(
+        qualifier: Option<&str>,
+        name: &str,
+        data_type: DataType,
+        nullable: bool,
+    ) -> Self {
+        DFField {
+            qualifier: qualifier.map(|s| s.to_owned()),
+            field: Field::new(name, data_type, nullable),
+        }
+    }
+
+    /// Create an unqualified field from an existing Arrow field
+    pub fn from(field: Field) -> Self {
+        Self {
+            qualifier: None,
+            field,
+        }
+    }
+
+    /// Create a qualified field from an existing Arrow field
+    pub fn from_qualified(qualifier: &str, field: Field) -> Self {
+        Self {
+            qualifier: Some(qualifier.to_owned()),
+            field,
+        }
+    }
+
+    /// Returns an immutable reference to the `DFField`'s unqualified name
+    pub fn name(&self) -> &String {
+        &self.field.name()
+    }
+
+    /// Returns an immutable reference to the `DFField`'s data-type
+    pub fn data_type(&self) -> &DataType {
+        &self.field.data_type()
+    }
+
+    /// Indicates whether this `DFField` supports null values
+    pub fn is_nullable(&self) -> bool {
+        self.field.is_nullable()
+    }
+
+    /// Returns an immutable reference to the `DFField`'s qualified name
+    pub fn qualified_name(&self) -> String {
+        if let Some(relation_name) = &self.qualifier {
+            format!("{}.{}", relation_name, self.field.name())
+        } else {
+            self.field.name().to_owned()
+        }
+    }
+
+    /// Get the optional qualifier
+    pub fn qualifier(&self) -> &Option<String> {
+        &self.qualifier
+    }
```

Comment:
```suggestion
    pub fn qualifier(&self) -> Option<&String> {
        self.qualifier.as_ref()
    }
```

I have been told that pattern is more idomatic rust, though the &Option works just fine

Reply:
Makes sense, thanks.